### PR TITLE
haskell.compiler.ghcHEAD: 9.3.20220406 -> 9.5.20220921

### LINKED ISSUE #137127: GHC: build with hadrian
I've given this a go, but it doesn't seem hadrian is packaged with Nix yet.

In theory, it should speed up the builds, but that's yet to be benchmarked.

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/haskell-packages.nix [hunk 1]
```diff
@@ -42,7 +44,10 @@ let
       enableExecutableProfiling = false;
       enableLibraryProfiling = false;
       enableSharedExecutables = false;
-      enableSharedLibraries = false;
+      # Usually necessary to build e.g. QuickCheck (required by hadrian).
+      # This binary GHC built on Alpine is statically linked and can't produce
+      # shared libraries.
+      enableSharedLibraries = self.ghc.name != "ghc-binary-musl-9.2.2";
```
Comment (sternenseemann) [inline pkgs/top-level/haskell-packages.nix:50]: Good news, we can likely drop this when [!8699](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8699) is merged and backported to `ghc-9.4`.

**Comment by sternenseemann** [comment]
* Native Compiler
  * [x] x86_64-linux
  * [x] x86_64-unknown-linux-mul
  * [x] aarch64-linux
  * [ ] x86_64-darwin
  * [ ] aarch64-darwin
  * [ ] (powerpc64le-linux)
* [ ] Cross compiler
  * [ ] -> LLVM linux
  * [ ] -> Native linux
  * [ ]  -> powerpc
  * [ ] -> Windows
* [ ] Static Cross Compiler

**Comment by TristanCacqueray** [comment]
Out of curiosity, is this the current fix for:
* #137127 

Thanks!

**Comment by sternenseemann** [comment]
Yes, but only for GHC HEAD. After evaluating the situation it does not seem worth it to port any GHC expression before 9.5 to hadrian, some of the features we need are broken and won't be fixed, there is no performance improvement for our use case.

**Pull request merged**